lib/location.py

```diff
@@ -90,106 +90,113 @@ def geolocator(location):
+    """
+    cachefile = os.path.join(IP2LCACHE, ip_addr)
     if not os.path.exists(IP2LCACHE):
         os.makedirs(IP2LCACHE)
 
-    location = None
-
-    if os.path.exists(cached):
-        location = open(cached, 'r').read().split(';')
-        if len(location) > 3:
-            return location[3], location[1]
-        elif len(location) > 1:
-            return location[0], location[1]
-        else:
-            return location[0], None
+    if os.path.exists(cachefile):
+        try:
+            city, region, country = open(cachefile, 'r').read().split(';')
+            return city, region, country
+        except ValueError:
+            # cache entry is malformed: should be city;region;country
+            return None
+    return None
 
-    return None, None
 
 def ip2location(ip_addr):
-    "Convert IP address `ip_addr` to a location name"
-
-    location = ipcache(ip_addr)
-    if location:
-        return location
-
-    # if IP2LOCATION_KEY is not set, do not the query,
+    """Convert IP address `ip_addr` to a location name"""
+    # if IP2LOCATION_KEY is not set, do not query,
     # because the query wont be processed anyway
-    if IP2LOCATION_KEY:
-        try:
-            location = requests\
-                    .get('http://api.ip2location.com/?ip=%s&key=%s&package=WS3' \
-                         % (ip_addr, IP2LOCATION_KEY)).text
-        except requests.exceptions.ConnectionError:
-            pass
-
-    if location and ';' in location:
-        ipcachewrite(ip_addr, location)
-        location = location.split(';')[3], location.split(';')[1]
-    else:
-        location = location, None
-
-    return location
+    if not IP2LOCATION_KEY:
+        return None, None, None
+    try:
+        r = requests.get(
+            'http://api.ip2location.com/?ip=%s&key=%s&package=WS3'
+            % (ip_addr, IP2LOCATION_KEY))
+        r.raise_for_status()
+        location = r.text
+        if location and ';' in location:
+            _, country, region, city = location.split(';')
+            location = city, region, country
+    except requests.exceptions.RequestException:
+        return None, None, None
+    return city, region, country
 
 
 def ipinfo(ip_addr):
-    location = ipcache(ip_addr)
-    if location:
-        return location
-    if IPINFO_TOKEN:
-        r = requests.get('https://ipinfo.io/%s/json?token=%s' %
-                         (ip_addr, IPINFO_TOKEN))
-        if r.status_code == 200:
-            location = r.json()["city"], r.json()["country"]
-    if location:
-        ipcachewrite(ip_addr, location)
-    return location
+    if not IPINFO_TOKEN:
+        return None, None, None
+    try:
+        r = requests.get(
+            'https://ipinfo.io/%s/json?token=%s'
+            % (ip_addr, IPINFO_TOKEN))
+        r.raise_for_status()
+        r_json = r.json()
+        city, region, country = r_json["city"], r_json["region"], r_json["country"]
```

Comment:
I would add an empty field (or the real country code if ipinfo.io delivers one) as the 0 field,
for the compatibility reasons

Reply:
All iplocating functions now return the country code. `ipinfo` in particular only returned 2 character country codes, so I had to add `pycountry` to correct for that, getting the real country name.